lib/spans/base-span-streamer.js

```diff
@@ -0,0 +1,135 @@
+/*
+ * Copyright 2020 New Relic Corporation. All rights reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+'use strict'
+
+const logger = require('../logger').child({ component: 'base-span-streamer' })
+const NAMES = require('../metrics/names').INFINITE_TRACING
+const SPAN_DROP_MSG_INTERVAL_MS = 30000
+const SPAN_DROP_MSG =
+  'Queue full, dropping spans. ' +
+  `Will not warn again for ${SPAN_DROP_MSG_INTERVAL_MS / 1000} seconds.`
+
+class BaseSpanStreamer {
+  constructor(licenseKey, connection, metrics, queueSize) {
+    this.stream = null
+    this.license_key = licenseKey
+    this.connection = connection
+    this.queue_size = queueSize
+    this.spans = []
+    this._metrics = metrics
+    this._writable = false
+
+    // 'connected' indicates a safely writeable stream.
+    // May still be mid-connect to gRPC server.
+    this.connection.on('connected', (stream) => {
+      logger.info('Span streamer connected')
+      this.stream = stream
+      this._writable = true
+      this.sendQueue()
+    })
+
+    this.connection.on('disconnected', () => {
+      logger.info('Span streamer disconnected')
+      this.stream = null
+      this._writable = false
+    })
+  }
+
+  addToQueue() {
+    throw new Error('addToQueue is not implemented')
+  }
+
+  /* Accepts a span and either writes it to the stream, queues it to be sent,
+   * or drops it depending on stream/queue state
+   */
+  write(span) {
+    this._metrics.getOrCreateMetric(NAMES.SEEN).incrementCallCount()
+
+    // If not writeable (because of backpressure) queue the span
+    if (!this._writable) {
+      if (this.spans.length < this.queue_size) {
+        this.addToQueue(span)
+        return
+      }
+
+      // While this can be directionally calculated between seen/sent the
+      // queue makes that a bit more disconnected. This will be a bit more specific.
+      this._metrics.getOrCreateMetric(NAMES.DROPPED).incrementCallCount()
+
+      // If the queue is full drop the span
+      logger.infoOncePer(
+        'SPAN_DROP_MSG', // key for the OncePer
```

Comment:
We probably should come up with some consistent naming scheme for these `oncePer` identifiers. We use them very infrequently, so it doesn't matter much, but so far we've used `'Spaces and caps'`, `'prefix:kebab-case'` and `'SHOUTY_SNAKE_CASE'`.

Reply:
I'm not sure I follow. are you suggesting `SPAN_DROP_MSG` is inconsistent? Either way this code is not new it just appears that way because it was moved

Reply:
I'm just saying we've been using all sorts of wacky identifiers for the `once`/`oncePer` logging statements, it looks funny here.

Reply:
I hear ya. I'm not going to address consistency across these messages because it's all over the agent